flang/lib/Lower/FIRBuilder.cpp

```diff
@@ -407,23 +408,29 @@ struct CharacterOpsBuilderImpl {
                       "Incorrect number of bounds in substring");
       return {mlir::Value{}, mlir::Value{}};
     }
-    auto indexType = mlir::IndexType::get(builder.getContext());
-    auto lowerBound = builder.createHere<fir::ConvertOp>(indexType, bounds[0]);
+    mlir::SmallVector<mlir::Value, 2> typedBounds;
+    // Convert bounds to length type to do safe arithmetic on it.
+    for (auto bound : bounds)
+      typedBounds.push_back(
+          builder.createHere<fir::ConvertOp>(builder.getLengthType(), bound));
+    auto lowerBound = typedBounds[0];
     // FIR CoordinateOp is zero based but Fortran substring are one based.
-    auto oneIndex = builder.createIntegerConstant(indexType, 1);
-    auto offsetIndex =
-        builder.createHere<mlir::SubIOp>(lowerBound, oneIndex).getResult();
+    auto one = builder.createIntegerConstant(lowerBound.getType(), 1);
+    auto offset = builder.createHere<mlir::SubIOp>(lowerBound, one).getResult();
+    auto idxType = builder.getIndexType();
+    if (offset.getType() != idxType)
+      offset = builder.createHere<fir::ConvertOp>(idxType, offset);
     auto substringRef = builder.createHere<fir::CoordinateOp>(
-        str.getReferenceType(), str.data, offsetIndex);
+        str.getReferenceType(), str.data, offset);
 
     // Compute the length.
     mlir::Value substringLen{};
     if (nbounds < 2) {
-      substringLen = builder.createHere<mlir::SubIOp>(str.len, bounds[0]);
+      substringLen = builder.createHere<mlir::SubIOp>(str.len, typedBounds[0]);
```

Comment:
All values are typed. This is probably suggesting something, but I'm not sure I'm following.

Reply:
The intention was to say that they have the required type to be used in the binary operations with a length, which is not safe to assume from bounds.  Renamed to `castBounds`.